src/MongoDB.Driver/ClusterKey.cs

```diff
@@ -171,13 +180,15 @@ public override bool Equals(object obj)
                 _heartbeatInterval == rhs._heartbeatInterval &&
                 _heartbeatTimeout == rhs._heartbeatTimeout &&
                 _ipv6 == rhs._ipv6 &&
+                GetValueOrEmpty(_kmsProviders).SequenceEqual(GetValueOrEmpty(rhs._kmsProviders), new KmsProvidersComparer()) &&
```

Comment:
GetValueOrEmpty creates a small amount of garbage every time Equals is called (if _kmsProviders is null).

You could implement a helper method that doesn't create garbage.